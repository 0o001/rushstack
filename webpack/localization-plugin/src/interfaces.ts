// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @public
 */
export interface IPassthroughLocaleOptions {
  usePassthroughLocale?: boolean;

  /**
   * If {@link IPassthroughLocaleOptions.usePassthroughLocale} is set, use this name for the passthrough locale.
   * Defaults to "passthrough"
   */
  passthroughLocaleName?: string;
}

/**
 * @public
 */
export interface ITypingsGenerationOptions {
  generatedTsFolder: string;
  sourceRoot?: string;
  exportAsDefault?: boolean;
}

/**
 * @public
 */
export interface IDefaultLocaleOptions {
  localeName: string;

  fillMissingTranslationStrings?: boolean;
}

/**
 * Options for the pseudolocale library.
 *
 * @internalRemarks
 * Eventually this should be replaced with DefinitelyTyped types.
 *
 * @public
 */
export interface IPseudolocaleOptions {
  prepend?: string;
  append?: string;
  delimiter?: string;
  startDelimiter?: string;
  endDelimiter?: string;
  extend?: number;
  override?: string;
}

/**
 * @public
 */
export interface IPseudolocalesOptions {
  [pseudoLocaleName: string]: IPseudolocaleOptions;
}

/**
 * @public
 */
export interface ILocalizedData {
  /**
   * Use this parameter to specify the translated data.
   */
  translatedStrings: ILocalizedStrings;

  /**
   * Options around including a passthrough locale.
   */
  passthroughLocale?: IPassthroughLocaleOptions;

  /**
   * Options for the locale used in the source localized data files.
   */
  defaultLocale: IDefaultLocaleOptions;

  /**
   * Options for pseudo-localization.
   */
  pseudolocales?: IPseudolocalesOptions;
}

/**
 * @public
 */
export interface ILocalizationStatsOptions {
  dropPath?: string;

  callback?: (stats: ILocalizationStats) => void;
}

/**
 * The options for localization.
 *
 * @public
 */
export interface ILocalizationPluginOptions {
  localizedData: ILocalizedData;

  filesToIgnore?: string[];

  localizationStats?: ILocalizationStatsOptions;

  typingsOptions?: ITypingsGenerationOptions;
}

/**
 * @internal
 */
export interface ILocalizationFile {
  [stringName: string]: ILocalizedString;
}

/**
 * @internal
 */
export interface ILocalizedString {
  value: string;
  comment?: string;
}

/**
 * @public
 */
export interface ILocaleFileData {
  [stringName: string]: string;
}

/**
 * @public
 */
export interface ILocaleData {
  [locFilePath: string]: ILocaleFileData;
}

/**
 * @public
 */
export interface ILocalizedStrings {
  [locale: string]: ILocaleData;
}


/**
 * @public
 */
export interface ILocalizationStatsEntrypoint {
  localizedAssets: { [locale: string]: string };
}

/**
 * @public
 */
export interface ILocalizationStatsChunkGroup {
  localizedAssets: { [locale: string]: string };
}

/**
 * @public
 */
export interface ILocalizationStats {
  entrypoints: { [name: string]: ILocalizationStatsEntrypoint };
  namedChunkGroups: { [name: string]: ILocalizationStatsChunkGroup };
}
