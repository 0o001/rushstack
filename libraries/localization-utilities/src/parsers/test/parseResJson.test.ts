// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { IgnoreStringFunction } from '../../interfaces';
import { parseResJson } from '../parseResJson';
import { serializeLocFile } from './serializeLocFile';

describe(parseResJson.name, () => {
  it('parses a valid file', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_foo.comment': 'A string',
      bar: 'Bar',
      '_bar.comment': 'Another string'
    });

    expect(
      serializeLocFile(
        parseResJson({
          content,
          filePath: 'test.resjson'
        })
      )
    ).toMatchInlineSnapshot(`
      "key | value | comment
      ------------------------------
       foo | Foo   | A string
       bar | Bar   | Another string"
    `);
  });

  it('throws on excess comments', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_bar.comment': 'A string'
    });

    expect(() =>
      parseResJson({
        content,
        filePath: 'test.resjson'
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('correctly ignores a string', () => {
    const content: string = JSON.stringify({
      foo: 'Foo',
      '_foo.comment': 'A string',
      bar: 'Bar',
      '_bar.comment': 'Another string'
    });

    const ignoredStringFunction: IgnoreStringFunction = jest
      .fn()
      .mockImplementation(
        (fileName: string, stringName: string) => fileName === 'test.resjson' && stringName === 'bar'
      );

    expect(
      serializeLocFile(
        parseResJson({
          content,
          filePath: 'test.resjson',
          ignoreString: ignoredStringFunction as IgnoreStringFunction
        })
      )
    ).toMatchInlineSnapshot(`
      "key | value | comment
      ------------------------
       foo | Foo   | A string"
    `);

    expect((ignoredStringFunction as unknown as jest.SpyInstance).mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "test.resjson",
          "foo",
        ],
        Array [
          "test.resjson",
          "bar",
        ],
      ]
    `);
  });
});
