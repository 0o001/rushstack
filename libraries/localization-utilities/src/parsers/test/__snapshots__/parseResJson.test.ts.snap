// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResJson throws on excess comments 1`] = `"The resjson file is invalid. Comments exist for the following string keys that don't have values: bar."`;
